features/nfc/nfc/NFCEEPROMDriver.h

```diff
@@ -0,0 +1,188 @@
+         */ 
+        void set_delegate(Delegate* delegate);
+
+        /** 
+         * Reset and initialize the EEPROM.
+         * This method should complete synchronously.
+         */ 
+        virtual void reset() = 0;
+
+        /**
+         * Get the maximum memory size addressable by the EEPROM.
+         */ 
+        virtual size_t get_max_size() = 0;
+
+        /**
+         * Start a session of operations (reads, writes, erases, size gets/sets).
+         * This method is called prior to any memory access to allow the underlying implementation 
+         * to disable the RF interface or abort the transaction if it's being used. 
+         * This method should complete asynchronously by calling has_started_session().
+         */ 
+        virtual void start_session() = 0; // This could lock the chip's RF interface
+
+        /**
+         * End a session.
+         * This method should complete asynchronously by calling has_ended_session().
+         */ 
+        virtual void end_session() = 0;
+
+        /**
+         * Read bytes from memory.
+         * @param[in] address the virtual address (starting from 0) from which to start the read.
+         * @param[out] bytes a buffer in which the read bytes will be stored.
+         *  This buffer should remain valid till the callback is called.
+         * @oaram[in] count the number of bytes to read.
+         * This method should complete asynchronously by calling has_read_bytes().
+         */ 
+        virtual void read_bytes(uint32_t address, uint8_t* bytes, size_t count) = 0;
+
+        /**
+         * Write bytes to memory.
+         * @param[in] address the virtual address (starting from 0) from which to start the write.
+         * @param[in] bytes a buffer from to copy.
+         *  This buffer should remain valid till the callback is called.
+         * @oaram[in] count the number of bytes to write.
+         * This method should complete asynchronously by calling has_written_bytes().
+         */ 
+        virtual void write_bytes(uint32_t address, const uint8_t* bytes, size_t count) = 0;
+
+        /**
+         * Set the size of the addressable memory.
+         * @oaram[in] count the number of addressable bytes.
+         * This method should complete asynchronously by calling has_set_size().
+         */ 
+        virtual void set_size(size_t count) = 0;
+
+        /**
+         * Get the size of the addressable memory.
+         * This method should complete asynchronously by calling has_gotten_size().
+         */ 
+        virtual void get_size() = 0;
+
+        /**
+         * Erase bytes from memory.
+         * @param[in] address the virtual address (starting from 0) from which to start erasing.
+         * @oaram[in] count the number of bytes to erase.
+         * This method should complete asynchronously by calling has_erased_bytes().
+         */ 
+        virtual void erase_bytes(uint32_t address, size_t size) = 0;
+
```

Comment:
need one more:
void on_handle_events() in delegate
void handle_events() in driver

so that async handling can be put on the event queue by the user, just like we  already do with event handling elsewhere